### ISSUE TITLE
CUDA 9.1 planning ticket

### ISSUE THREAD

**Issue by tfboyd** [tensorflow/tensorflow#15140]
**Update 05-FEB-2018**
The plan of record is to stick with CUDA 9.x until CUDA 10.  That plan has an issue in that CUDA 9 and 9.1 cause problems with XLA that should be resolved with CUDA 9.2.  The soft plan is we would move to 9.2 when it comes out if it resolves the issues with XLA.  

**Update 23-JAN-2018**
CUDA 9.1 requires an upgrade to device driver 387 (CUDA 9 was 384).  Moving device drivers is painful for production environments.  We are not going to move the default builds to CUDA 9.1 or 9.2 we will stick with CUDA 9 likely until CUDA 10.  We will move cuDNN forward which will have a larger impact and not require device drive upgrades.  This space is developing as everyone involved evolves their processes and learns from the past.  

While I cannot promise anything, I do want to create a "channel" where we are building and testing the latest CUDA 9.x so we can track performance improvements and have some avenue for people to get those builds.  The testing infrastructure is large and maintaining this has a cost.  I hope to find a middle ground as I like perf testing the latest libraries.  

**Original Message**
The purpose of this thread is to keep CUDA 9.1 questions related to when it will be in TensorFlow in a single area.  Separate issues are fine.  I will try to keep this first comment updated with information as it comes out.  

**Current Status:**  Unknown, waiting for RC and gathering information to formulate a plan.  

p.s. There is a tendency to treat TensorFlow like a one way product.  I want to continue to change that with this type of dialog and transparency.  Many people outside Google will contribute to CUDA 9.1 support for TensorFlow.

**Comment by Codewyvern** [comment]
Is cuda 9.1 an upcoming release? I can only download cuda_9.0.176 from the nvidia dev portal. Will this version be supported in the RC too?

**Comment by sclarkson** [comment]
CUDA 9.1 has been released.

Some hpp headers have been removed, namely math_functions.hpp which is included by Eigen/Core.

Noting that math_functions.h just diverts to crt/math_functions.h, I tried the same with the hpp version. I put this into /usr/local/cuda/math_functions.hpp.
```
#include "crt/math_functions.hpp"
```

Compilation and a basic test of 1.4.0 worked. Didn't run the bazel tests.

**Comment by tfboyd** [comment]
CUDA 9.1 will require driver 387+.  This means enterprises have to upgrade their fleets.  I have not looked at a timeline for 9.1.  It should work in source (someone internally was using it the other day but I did not ask how or even with TensorFlow directly) now but maybe not using all the latest optimized features that were added.  We will likely not look to closely until Jan.  Once I get things together I will try to publish a very unofficial timeline.  If anyone has verified it works (even just with CUDA 9 level perf) at head let me know.  I am too busy to try right now.

**Comment by sclarkson** [comment]
Running https://github.com/tensorflow/benchmarks/commit/0f0492dd5590549e1ce7d1d91bf5c88794f95f69 (a few commits after this introduces an operation from master that isn't in 1.4).

```
python tf_cnn_benchmarks.py --num_gpus=4 --batch_size=64 --model=inception3 --variable_update=parameter_server --local_parameter_device=cpu
```

Machine is 4x Titan V with driver 387.3. Both versions are Tensorflow 1.4.1.

| GPUs | CUDA 9.0.176 + cuDNN 7.0.5 | CUDA 9.1.85 + cuDNN 7.0.5 |
| :---: | :---: | :---: |
| 1 | 191.56 images/sec | 191.59 images/sec |
| 4 | 727.90 images/sec | 738.54 images/sec |

I only ran these once, so I'm not sure that the increased 4 GPU performance is significant. I can run it on imagenet data if you'd like to see that, as well.

**Comment by gdshen** [comment]
copy /usr/local/cuda/include/math_functions.h to /usr/local/include/math_functions.hpp, and it works.
```bash
sudo cp /usr/local/cuda/include/math_functions.h /usr/local/cuda/include/math_functions.hpp
```

**Comment by domschl** [comment]
@gdshen : That's a bit dangerous, because you might end up with an outdated copy on next cuda-update, if you forget about that second copy.
@sclarkson 's approach is somewhat safer. This is anyway most probably a CUDA-9.1 issue.

**Comment by gdshen** [comment]
@domschl A soft copy may be safer. 
It would be more elegant to put header file in /usr/local/cuda/include, not in /usr/local/cuda

**Comment by bruce25796** [comment]
cd /usr/local/cuda/include

There are two solutions:
(A)
sudo ln -s math_functions.h math_functions.hpp

(B)
sudo ln  -s  /usr/local/cuda/include/crt/math_functions.hpp math_functions.hpp 

I think choice(B) might also works.

**Comment by tfboyd** [comment]
@sclarkson   You made me jealous.  I do not have a good way to have 4 Titan Vs but I ordered two for the t team so we could help ensure things are working when the community has issues.  Things will slow down for the holidays but we have discussed 9.1 and the team desires to roll it out fast in Q1 as well as finish up the CUDA 9 upgrade.

**Comment by svenstaro** [comment]
A better way to make this compile in the meantime is to just add the cuda crt include path to the cxxopts as I have done in the Arch Linux package: https://git.archlinux.org/svntogit/community.git/commit/trunk/PKGBUILD?h=packages/tensorflow&id=530561e9e6bac127d1091f206f6214c967072833

**Comment by leolle** [comment]
Hi, I built the tensorflow 1.4.1 sauce code on cuda9.1, but I got the error message like this:
`ImportError: libcublas.so.8.0: cannot open shared object file: No such file or directory
`
can anybody help?

**Comment by bruce25796** [comment]
Tensorflow 1.5 rc  is built  under cuda 9.1 and cudnn 7.0 successfully.

**Comment by arunmandal53** [comment]
Go to http://www.python36.com/install-tensorflow141-gpu/ for step by step installation of tensorflow with cuda 9.1 and cudnn7.05 on ubuntu.

**Comment by Vimos** [comment]
Using Plan(C) on r1.4 as svenstaro suggested, the compiling still complains:
```
ERROR: ~/git/tensorflow/tensorflow/tensorflow/contrib/rnn/BUILD:193:1: Couldn't build file tensorflow/contrib/rnn/_objs/python/ops/_lstm_ops_gpu/tensorflow/contrib/rnn/kernels/lstm_ops_gpu.cu.pic.o: undeclared inclusion(s) in rule '//tensorflow/contrib/rnn:python/ops/_lstm_ops_gpu':
this rule is missing dependency declarations for the following files included by 'tensorflow/contrib/rnn/kernels/lstm_ops_gpu.cu.cc':
  '/usr/local/cuda/include/crt/math_functions.hpp'
  '/usr/local/cuda/include/crt/host_defines.h'
```

**Comment by Vimos** [comment]
Using Plan(C) , 
```
bazel build -c opt --copt=-mavx --copt=-mavx2 --copt=-mfma --copt=-mfpmath=both --copt=-msse4.2 --config=cuda --incompatible_load_argument_is_label=false --cxxopt="-I/usr/local/cuda/include/crt" -k //tensorflow/tools/pip_package:build_pip_package
```

the compiling complains:
```
ERROR: /home/vimos/Downloads/tensorflow-1.4.1/tensorflow/contrib/resampler/BUILD:45:1: Couldn't build file tensorflow/contrib/resampler/_objs/python/ops/_resampler_ops_gpu/tensorflow/contrib/resampler/kernels/resampler_ops_gpu.cu.pic.o: undeclared inclusion(s) in rule '//tensorflow/contrib/resampler:python/ops/_resampler_ops_gpu':
this rule is missing dependency declarations for the following files included by 'tensorflow/contrib/resampler/kernels/resampler_ops_gpu.cu.cc':
  '/usr/local/cuda/include/crt/math_functions.hpp'
  '/usr/local/cuda/include/crt/host_defines.h'

```

**Comment by yqm0nk3y** [comment]
Just wondering if anyone managed to build CUDA9.1 with tensorflow on windows 10 ? Thanks

**Comment by arunmandal53** [comment]
Yes. I have built it. just testing it. Tensorflow gpu with cuda 9.1 and cudnn 7.0.5. I have built it with bazel and cmake. But using cmake is recommended by contributers.

**Comment by arunmandal53** [comment]
Finally created a working tutorial for building Tensorflow 1.5.0 on windows with cuda 9.1 and cudnn 7.0.5 using cmake and visual studio 2015 update 3. I have tested and worked. Please comment if it worked for you too. Here is the link for blog http://www.python36.com/install-tensorflow-gpu-windows

**Comment by Hyperclaw79** [comment]
Any plans to include support for VS 2017 too?

**Comment by ultrasounder** [comment]
I followed the instructions verbatim and I get this - After 11hrs and 18 mins of build. 
![tensorflow_build_source_errors](https://user-images.githubusercontent.com/2142626/35197294-0e94ab78-fe92-11e7-9d8e-83fbc32e8b99.png)

**Comment by rolla12** [comment]
arunmandal53c can you update this http://www.python36.com/install-tensorflow-gpu-windows to include bazel too

**Comment by arunmandal53** [comment]
@rolla12 Using cmake is more stable than bazel on windows. I will update the tutorial if issues in bazel build fixed for windows. Using Network installer of cuda 9.1 toolkit update your driver to latest version which may have compatibility with latest linux kernel. If you have issue to load nvidia driver then you can downgrade linux kernel by going to advanced option for ubuntu in boot menu. Then reinstalling cuda toolkit 9.1 will fix the driver issue. Currently on my test linux kernel 4.10 worked.

**Comment by rolla12** [comment]
use it worked with 4.10 but give not sufficient driver with 4.13

**Comment by bringtree** [comment]
i build it successfully.However, when i import it. it throw an error.
`bringtree@bringtree-HP-Z420-Workstation:~/tensorflow-1.5.0/tensorflow_pkg$ python3
Python 3.5.2 (default, Nov 23 2017, 16:37:01)
[GCC 5.4.0 20160609] on linux
Type "help", "copyright", "credits" or "license" for more information.
>>> import tensorflow as tf
t>>> sess= tf.Session()
2018-01-28 20:14:56.074102: I tensorflow/core/common_runtime/gpu/gpu_device.cc:1105] Found device 0 with properties:
name: GeForce GTX 1060 6GB major: 6 minor: 1 memoryClockRate(GHz): 1.8475
pciBusID: 0000:05:00.0
totalMemory: 5.93GiB freeMemory: 5.47GiB
2018-01-28 20:14:56.074141: I tensorflow/core/common_runtime/gpu/gpu_device.cc:1195] Creating TensorFlow device (/device:GPU:0) -> (device: 0, name: GeForce GTX 1060 6GB, pci bus id: 0000:05:00.0, compute capability: 6.1)
2018-01-28 20:14:56.074277: E tensorflow/core/common_runtime/direct_session.cc:168] Internal: cudaGetDevice() failed. Status: CUDA driver version is insufficient for CUDA runtime version
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
  File "/usr/local/lib/python3.5/dist-packages/tensorflow/python/client/session.py", line 1509, in __init__
    super(Session, self).__init__(target, graph, config=config)
  File "/usr/local/lib/python3.5/dist-packages/tensorflow/python/client/session.py", line 628, in __init__
    self._session = tf_session.TF_NewDeprecatedSession(opts, status)
  File "/usr/local/lib/python3.5/dist-packages/tensorflow/python/framework/errors_impl.py", line 473, in __exit__
    c_api.TF_GetCode(self.status.status))
tensorflow.python.framework.errors_impl.InternalError: Failed to create session.`

**Comment by arunmandal53** [comment]
@bringtree what the command `nvidia-smi` returns?

**Comment by tfboyd** [comment]
@bringtree Your likely problem bsed on the error

> CUDA driver version is insufficient for CUDA runtime version

- CUDA 9 needs driver 384.x+
- CUDA 9.1 needs driver 387.x+

I am 99% sure if I am wrong I will update this or delete if someone has a cleaner comment.

**Comment by rolla12** [comment]
for cuda 9.1 the driver is 390 ,, its know included with deb installation ,,it solved the problem

**Comment by arunmandal53** [comment]
@ultrasounder pip install wheel should fix the problem before build. Sorry for late reply.

**Comment by dchichkov** [comment]
Mind that Linux 4.13.x (default on 16.04.3) has some issues with 387.22+.
 https://bugs.launchpad.net/ubuntu/+source/linux/+bug/1742302

**Comment by rolla12** [comment]
@dchichkov  that issues has been solved by the driver 390

**Comment by tensorflowbutler** [comment]
Nagging Assignee: It has been 14 days with no activity and this issue has an assignee. Please update the label and/or status accordingly.

**Comment by ViktorM** [comment]
Hi! What were a reasons why TF 1.6 hasn't been shipped with CUDA 9.1 but only 9.0?

**Comment by tensorflowbutler** [comment]
Nagging Assignee @tfboyd: It has been 14 days with no activity and this issue has an assignee. Please update the label and/or status accordingly.

**Comment by tensorflowbutler** [comment]
Nagging Assignee @tfboyd: It has been 14 days with no activity and this issue has an assignee. Please update the label and/or status accordingly.

**Comment by tensorflowbutler** [comment]
Nagging Assignee @tfboyd: It has been 14 days with no activity and this issue has an assignee. Please update the label and/or status accordingly.

**Comment by ViktorM** [comment]
And CUDA 9.2 is out. Are there any updates on what TF version is going to be shipped with 9.2 and what with 10.0.

**Comment by IoanaPo** [comment]
Hi. I am trying to run tensorflow-gpu (1.7) on Windows 10, with CUDA 9.1 and cudnn 7.0.5, python 3.6 and a GeForce 1070, and I am getting a cublas error: "failed to create cublas handle: CUBLAS_STATUS_ALLOC_FAILED". I've Googled the error and I've tried the few solutions mentioned there (calling the GPU before the CPU and not vice-versa, trying to reinstall things, etc), but nothing worked. Does anyone know why I am getting this error and if there is a fix?

I followed the instructions  in this link http://www.python36.com/install-tensorflow-gpu-windows/ to install Tensorflow, but instead of building it myself, I used the prebuilt wheel from here https://drive.google.com/drive/folders/1lVK_ABvVHzVYKs7X5SUhcZFBgKpC41Qw
